pandas/core/series.py

```diff
@@ -162,21 +164,27 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
             elif isinstance(data, dict):
                 if index is None:
                     if isinstance(data, OrderedDict):
-                        index = Index(data)
+                        original_index = data.keys()
                     else:
-                        index = Index(_try_sort(data))
+                        original_index = _try_sort(data)
+
+                    index = Index(original_index)
+
                 try:
-                    if isinstance(index, DatetimeIndex):
-                        # coerce back to datetime objects for lookup
-                        data = lib.fast_multiget(data, index.astype('O'),
-                                                 default=np.nan)
-                    elif isinstance(index, PeriodIndex):
-                        data = [data.get(i, nan) for i in index]
+                    # lib.fast_multiget raises TypeError if type(data) != dict
+
+                    if lib.infer_dtype(data) == lib.infer_dtype(index.values):
+                        data = lib.fast_multiget(data, index.values, default=np.nan)
                     else:
-                        data = lib.fast_multiget(data, index.values,
-                                                 default=np.nan)
-                except TypeError:
-                    data = [data.get(i, nan) for i in index]
+                        if isinstance(original_index, PeriodIndex):
+                            data = [data.get(i, np.nan) for i in original_index]
+                        else:
+                            # np.array(['z', ('a', 'b')]) raises ValueError;
+                            # this may happens with MultiIndex.
+                            data = lib.fast_multiget(data, np.array(original_index),
```

Comment:
If we `lib.fast_multiget` supported `list` as its second parameter, we could save the call to `np.array`.